### PR TITLE
Mention foldk, add zip with example and add while

### PR DESCRIPTION
**Pull request by hgsipiere**

### CODE REVIEW

**Comment by hgsipiere** [comment]
The second commit is to address issues #39 and #45 .

File: docs/source/scilla-in-depth.rst [hunk 1]
```diff
@@ -940,6 +940,29 @@ can be used to traverse all the elements of any list:
...
+    accumulator (of type ``'B``). The second argument is the next list
+    element to be processed (of type ``'A``). The third argument is
+    the recursive call (of type ``'B -> 'B``). The result of the
+    function is the next value of the accumulator (of type ``'B``).
+
+  - The initial value of the accumulator ``z`` (of type ``'B``).
+
+  - The list of elements to be processed (of type ``List 'A``).
+
+  In the case of a non-empty last argument with head ``x`` and tail ``xs``
```
Comment (anton-trunov) [inline docs/source/scilla-in-depth.rst]: I imagine this paragraph might sound a bit confusing for newcomers. Let's remove it and show how to use the third argument of the combining function by example.

Comment (hgsipiere) [reply docs/source/scilla-in-depth.rst]: Should we do the same for nat_foldk?

Comment (anton-trunov) [reply docs/source/scilla-in-depth.rst]: yes please

Comment (hgsipiere) [reply docs/source/scilla-in-depth.rst]: Okay I'll explain nat_foldk through nat_eq, and just cut out lines 943 to 965.

File: docs/source/scilla-in-depth.rst [hunk 2]
```diff
@@ -1112,7 +1133,63 @@ which can be used to traverse all the Peano numbers from a given
   (of type ``'A``) when all Peano numbers down to ``Zero`` have been
   processed.
 
+- ``nat_foldk: ('A -> Nat -> ('A -> 'A) -> 'A) -> 'A -> Nat -> 'A``:
+  See below for an example and explanation.
```
Comment (anton-trunov) [inline docs/source/scilla-in-depth.rst]: Can we please have a high-level description here? Like the one we have for `nat_fold` and the list folds.

File: docs/source/scilla-in-depth.rst [hunk 3]
```diff
@@ -1112,7 +1133,63 @@ which can be used to traverse all the Peano numbers from a given
...
+    end
+
+Line 2 specifies that we take two Peano numbers ``m`` and ``n``.
+Line 3 instantiates the type of ``nat_foldk``, we give it ``Nat``
+because we will be passing a ``Nat`` value as the fold accumulator.
+
+Lines 4 to 8 specify the fold description, this is the first argument
+that ``nat_foldk`` takes usually of type ``'A -> Nat -> ('A -> 'A) -> 'A``
+but we have specified that ``'A`` is ``Nat`` in this case. Our function
+takes the accumulator ``n`` and ``ignore : Nat`` is the Peano number to
```
Comment (anton-trunov) [inline docs/source/scilla-in-depth.rst]: we should explain `ignore`

File: docs/source/scilla-in-depth.rst [hunk 4]
```diff
@@ -1112,7 +1133,63 @@ which can be used to traverse all the Peano numbers from a given
...
+be processed. In the case that ``ignore`` is zero ``recurse x`` returns
+``x``. In other case, let ``ignore_pred`` be the predecessor of ``ignore``
```
Comment (anton-trunov) [inline docs/source/scilla-in-depth.rst]: `ignore_pred` is missing. Could you revise the paragraph?

**Comment by anton-trunov** [comment]
@hgsipiere I tweaked the description of `list_foldk` a bit. Can we add explanations of how to express the right and left folds in terms of `list_foldk`? This should help our users to grasp how it works. I think that (and your `list_find` example) should be enough for this PR.

**Comment by hgsipiere** [comment]
Okay, I'll use the ones from https://github.com/Zilliqa/scilla/issues/611

**Pull request merged**